### PR TITLE
fix(Create3): checking for empty account with native ether balance

### PR DESCRIPTION
**Pull request by re1ro**
* [x] `Create3`: checking codehash for `keccak256('')`
* [x] `Upgradable`: better only proxy check
* [x] `Proxy` test: removed solidity warnings

### CODE REVIEW

File: contracts/upgradable/Upgradable.sol [hunk 1]
```diff
@@ -8,10 +8,15 @@ import { Ownable } from '../utils/Ownable.sol';
...
+    address internal immutable implementationAddress;
+
+    constructor() {
+        implementationAddress = address(this);
+    }
 
     modifier onlyProxy() {
         // Prevent setup from being called on the implementation
-        if (implementation() == address(0)) revert NotProxy();
+        if (address(this) == implementationAddress) revert NotProxy();
```
Comment (sammy1991106) [inline contracts/upgradable/Upgradable.sol:19]: Is there a unit test for this?

**Review by sammy1991106 — APPROVED** [review]

File: contracts/upgradable/Upgradable.sol [hunk 2]
```diff
@@ -8,10 +8,15 @@ import { Ownable } from '../utils/Ownable.sol';
...
```
Comment (deanamiel) [inline contracts/upgradable/Upgradable.sol:20]: This check is a bit confusing as the implementationAddress is immutable and is set to address(this) in the constructor, is it that address(this) when called in the constructor is actually the address of the proxy?

Comment (re1ro) [reply contracts/upgradable/Upgradable.sol:20]: When the implementation is called though `delegatecall` by the proxy. `address(this)` becomes the address of the proxy

Comment (deanamiel) [reply contracts/upgradable/Upgradable.sol:20]: Makes sense, so actually the opposite of what I initially thought

**Review by deanamiel — APPROVED** [review]

File: contracts/deploy/Create3.sol [hunk 3]
```diff
@@ -26,8 +26,11 @@ library Create3 {
 
     function deploy(bytes32 salt, bytes memory bytecode) internal returns (address deployed) {
         deployed = deployedAddress(salt, address(this));
+        bytes32 existingCodeHash = deployed.codehash;
 
-        if (deployed.codehash != bytes32(0)) revert AlreadyDeployed();
+        // https://eips.ethereum.org/EIPS/eip-1052
+        // keccak256('') == 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470
+        if (existingCodeHash != bytes32(0) && existingCodeHash != 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470) revert AlreadyDeployed();
```
Comment (milapsheth) [inline contracts/deploy/Create3.sol:33]: Can you add a unit test for this? To make this more reusable, maybe add a library method `isContract` and use it for `address` type? We'll likely need this again.

Comment (milapsheth) [inline contracts/deploy/Create3.sol:33]: This also needs to be handled during the sanity check at line 44. For e.g. a contract can `selfdestruct` and use it's own address to send the remaining funds, so it's codehash won't be 0. I'd also be open to remove the sanity check on line 44 altogether and rely on line 39 to check for failure.

File: contracts/upgradable/Upgradable.sol (hunk 1 above)
Comment (deanamiel) [reply contracts/upgradable/Upgradable.sol:19]: I'll add this test

File: contracts/deploy/Create3.sol (hunk 3 above)
Comment (deanamiel) [reply contracts/deploy/Create3.sol:33]: I can add tests, were you thinking of adding it to an existing library or should I create a new one?

Comment (deanamiel) [reply contracts/deploy/Create3.sol:33]: I can add the additional check to line 44

Comment (milapsheth) [reply contracts/deploy/Create3.sol:33]: New one, since existing libraries don't seem relevant to this

Comment (milapsheth) [reply contracts/deploy/Create3.sol:33]: Thanks

File: contracts/deploy/Create3.sol [hunk 4]
```diff
@@ -41,7 +53,7 @@ library Create3 {
         deployer.deploy(bytecode);
 
         // checking for codehash instead of code length to support contracts that selfdestruct in constructor
-        if (deployed.codehash == bytes32(0)) revert DeployFailed();
+        if (!deployed.isContract()) revert DeployFailed();
```
Comment (milapsheth) [inline contracts/deploy/Create3.sol:56]: Do we want to keep this sanity check?

File: contracts/deploy/Create3.sol [hunk 5]
```diff
@@ -21,13 +21,28 @@ contract CreateDeployer {
     }
 }
 
+library AddressUtils {
```
Comment (re1ro) [inline contracts/deploy/Create3.sol:24]: @deanamiel could you please move this into its own file under the `utils` folder? So we can also use it in other places as well. Let's call it `ContractAddress` to match the convention we have for other utils.

File: contracts/deploy/Create3.sol [hunk 6]
```diff
@@ -21,13 +21,28 @@ contract CreateDeployer {
...
 library Create3 {
+    using AddressUtils for address;
     bytes32 internal constant DEPLOYER_BYTECODE_HASH = keccak256(type(CreateDeployer).creationCode);
 
     function deploy(bytes32 salt, bytes memory bytecode) internal returns (address deployed) {
         deployed = deployedAddress(salt, address(this));
 
-        if (deployed.codehash != bytes32(0)) revert AlreadyDeployed();
+        // https://eips.ethereum.org/EIPS/eip-1052
+        // keccak256('') == 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470
```
Comment (re1ro) [inline contracts/deploy/Create3.sol:44]: I think we can remove this comment as `0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470` moved into the library and irrelevant in this context.

File: contracts/deploy/Create3.sol (hunk 4 above)
Comment (re1ro) [reply contracts/deploy/Create3.sol:56]: @deanamiel please remove this line as we already doing the success check here
https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/pull/32/files#diff-e8da8b3175680e091c0f1df471b0fda0ab20b884aa1dea000c4dd2caab59cdebL15-L17

Comment (deanamiel) [reply contracts/deploy/Create3.sol:56]: Sounds good

File: contracts/deploy/Create3.sol (hunk 6 above)
Comment (deanamiel) [reply contracts/deploy/Create3.sol:44]: Good point

File: contracts/deploy/Create3.sol [hunk 7]
```diff
@@ -22,12 +24,13 @@ contract CreateDeployer {
 }
 
 library Create3 {
+    using ContractAddress for address;
```
Comment (milapsheth) [inline contracts/deploy/Create3.sol:27]:
```suggestion
    using ContractAddress for address;

```
newline for readability

**Review by milapsheth — APPROVED** [review]

**Pull request merged**